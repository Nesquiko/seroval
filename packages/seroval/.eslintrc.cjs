module.exports = {
  "root": true,
  "extends": [
    'lxsmnsyc/typescript',
  ],
  "parserOptions": {
    "project": "./tsconfig.eslint.json",
    "tsconfigRootDir": __dirname,
  },
  "rules": {
    "import/no-extraneous-dependencies": [
      "error", {
        "devDependencies": ["**/*.test.ts"]
      }
    ],
    "prefer-template": "off",
    "@typescript-eslint/restrict-plus-operands": "off",
    "prefer-object-spread": "off",
    "class-methods-use-this": "off",
    "@typescript-eslint/no-explicit-any": "off"
  }
};