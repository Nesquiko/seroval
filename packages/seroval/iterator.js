function* test() {

}

function inheritance(value) {
  const p = Object.getPrototypeOf(value);
  if (p) {
    console.log(p);
    console.log(Object.getOwnPropertyDescriptors(p));
    inheritance(p);
  }
}

inheritance(test());
inheritance(new ReadableStream());