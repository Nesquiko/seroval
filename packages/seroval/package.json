{
  "name": "seroval",
  "type": "module",
  "version": "0.7.0",
  "files": [
    "dist",
    "src"
  ],
  "engines": {
    "node": ">=10"
  },
  "license": "MIT",
  "keywords": [
    "pridepack"
  ],
  "devDependencies": {
    "@types/node": "^18.15.3",
    "@vitest/ui": "^0.30.1",
    "eslint": "^8.32.0",
    "eslint-config-lxsmnsyc": "^0.6.3",
    "node-fetch-native": "^1.1.0",
    "pridepack": "2.4.1",
    "tslib": "^2.4.1",
    "typescript": "^4.9.4",
    "vitest": "^0.28.1"
  },
  "scripts": {
    "prepublishOnly": "pridepack clean && pridepack build",
    "build": "pridepack build",
    "type-check": "pridepack check",
    "lint": "pridepack lint",
    "clean": "pridepack clean",
    "watch": "pridepack watch",
    "start": "pridepack start",
    "dev": "pridepack dev",
    "test": "vitest",
    "test:ui": "vitest --ui"
  },
  "private": false,
  "description": "Stringify JS values",
  "repository": {
    "url": "https://github.com/lxsmnsyc/seroval.git",
    "type": "git"
  },
  "homepage": "https://github.com/lxsmnsyc/seroval/tree/main/packages/seroval",
  "bugs": {
    "url": "https://github.com/lxsmnsyc/seroval/issues"
  },
  "publishConfig": {
    "access": "public"
  },
  "author": "Alexis Munsayac",
  "types": "./dist/types/index.d.ts",
  "main": "./dist/cjs/production/index.cjs",
  "module": "./dist/esm/production/index.mjs",
  "exports": {
    ".": {
      "development": {
        "require": "./dist/cjs/development/index.cjs",
        "import": "./dist/esm/development/index.mjs"
      },
      "require": "./dist/cjs/production/index.cjs",
      "import": "./dist/esm/production/index.mjs",
      "types": "./dist/types/index.d.ts"
    }
  },
  "typesVersions": {
    "*": {}
  }
}
