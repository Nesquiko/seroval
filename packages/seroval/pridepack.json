{
  "target": "es2018",
  "entrypoints": {
    ".": "src/index.ts",
    "./plugins/web/url": "plugins/web/url/index.ts"
  }
}