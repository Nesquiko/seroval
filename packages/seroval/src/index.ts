import { forEach, join } from './array';
import getIdentifier from './get-identifier';
import quote from './quote';

const EMPTY_SET = 'new Set';
const EMPTY_MAP = 'new Map';
const EMPTY_ARRAY = '[]';

// Values that are non-recursive
export type PrimitiveValue =
  | boolean
  | string
  | number
  | undefined
  | null
  | bigint
  | Date
  | RegExp;

export type ServerValue =
  | PrimitiveValue
  | Array<ServerValue>
  | { [key: string | number]: ServerValue }
  | Set<ServerValue>
  | Map<ServerValue, ServerValue>;

// Reference counter
const REF_COUNT = new Map<ServerValue, number>();

const STACK = new Set();

// Value-to-ref map
const REFS = new Map<ServerValue, number>();

// Set to check if ref id already has an assigned value
const ASSIGNED_REFS = new Set<number>();

// Variables
let VARS: string[] = [];

// Array of assignments to be done (used for recursion)
type Assignment = [source: string, value: string];
let ASSIGNMENTS: Assignment[] = [];

// Array of Map.prototype.set calls
type MapSet = [source: string, key: string, value: string];
let MAP_SETS: MapSet[] = [];

// Array of Set.prototype.add calls
type SetAdd = [source: string, value: string];
let SET_ADDS: SetAdd[] = [];

function reset() {
  ASSIGNMENTS = [];
  MAP_SETS = [];
  SET_ADDS = [];
  VARS = [];
  REF_COUNT.clear();
  REFS.clear();
  ASSIGNED_REFS.clear();
}

function resolveAssignments() {
  if (ASSIGNMENTS.length) {
    const result: Record<string, string> = {};

    // Merge all assignments with similar source
    forEach(ASSIGNMENTS, ([source, value]) => {
      if (value in result) {
        result[value] = `${source}=${result[value]}`;
      } else {
        result[value] = `${source}=${value}`;
      }
    });

    return `${join(Object.values(result), ',')},`;
  }
  return '';
}

function resolveMapSets() {
  if (MAP_SETS.length) {
    const result: Record<string, string> = {};

    // Merge all assignments with similar source
    forEach(MAP_SETS, ([source, key, value]) => {
      if (source in result) {
        result[source] = `${result[source]}.set(${key},${value})`;
      } else {
        result[source] = `${source}.set(${key},${value})`;
      }
    });

    return `${join(Object.values(result), ',')},`;
  }
  return '';
}

function resolveSetAdds() {
  if (SET_ADDS.length) {
    const result: Record<string, string> = {};

    // Merge all assignments with similar source
    forEach(SET_ADDS, ([source, value]) => {
      if (source in result) {
        result[source] = `${result[source]}.add(${value})`;
      } else {
        result[source] = `${source}.add(${value})`;
      }
    });

    return `${join(Object.values(result), ',')},`;
  }
  return '';
}

function resolvePatches() {
  return `${resolveAssignments()}${resolveMapSets()}${resolveSetAdds()}`;
}

function insertRef(current: ServerValue) {
  const count = REF_COUNT.get(current) || 0;
  REF_COUNT.set(current, count + 1);
  return count === 0;
}

function constructorCheck<T extends NonNullable<ServerValue>>(
  value: NonNullable<ServerValue>,
  constructor: unknown,
): value is T {
  return value.constructor === constructor;
}

function isPrimitive(current: ServerValue): current is PrimitiveValue {
  if (!current || current === true) {
    return true;
  }
  const type = typeof current;
  const classConst = current.constructor;
  return type === 'number'
    || type === 'string'
    || type === 'bigint'
    || classConst === Date
    || classConst === RegExp;
}

function lookupRefs(current: ServerValue) {
  if (isPrimitive(current)) {
    return;
  }
  if (Array.isArray(current)) {
    if (insertRef(current)) {
      forEach(current, lookupRefs);
    }
  } else if (constructorCheck<Set<ServerValue>>(current, Set)) {
    if (insertRef(current)) {
      for (const item of current.keys()) {
        lookupRefs(item);
      }
    }
  } else if (constructorCheck<Map<ServerValue, ServerValue>>(current, Map)) {
    if (insertRef(current)) {
      for (const [key, value] of current.entries()) {
        lookupRefs(key);
        lookupRefs(value);
      }
    }
  } else if (current.constructor === Object && insertRef(current)) {
    forEach(Object.values(current), lookupRefs);
  }
}

function createRef(value: ServerValue) {
  const current = REFS.get(value);
  if (current != null) {
    return current;
  }
  const id = REFS.size;
  REFS.set(value, id);
  return id;
}

function getRefParam(index: number) {
  if (VARS[index]) {
    return VARS[index];
  }
  const result = getIdentifier(index);
  VARS[index] = result;
  return result;
}

function assignRef(index: number, value: string) {
  return `${getRefParam(index)}=${value}`;
}

function createAssignment(source: string, value: string) {
  ASSIGNMENTS.push([source, value]);
}

function createSetAdd(ref: number, value: string) {
  SET_ADDS.push([getRefParam(ref), value]);
}

function createMapSet(ref: number, key: string, value: string) {
  MAP_SETS.push([getRefParam(ref), key, value]);
}

function createArrayAssign(ref: number, index: number, value: string) {
  createAssignment(`${getRefParam(ref)}[${index}]`, value);
}

function createObjectIdentifierAssign(ref: number, key: string, value: string) {
  createAssignment(`${getRefParam(ref)}.${key}`, value);
}

function createObjectStringAssign(ref: number, key: string, value: string) {
  createAssignment(`${getRefParam(ref)}[${quote(key)}]`, value);
}

function serializePrimitive(value: PrimitiveValue): string {
  // Shortened forms
  if (value === true) {
    return '!0';
  }
  if (value === false) {
    return '!1';
  }
  if (value === undefined) {
    return 'void 0';
  }
  if (value === null) {
    return 'null';
  }
  // negative 0 isn't the same as 0
  if (Object.is(value, -0)) {
    return '-0';
  }
  if (typeof value === 'bigint') {
    return `BigInt("${value}")`;
  }
  if (typeof value === 'string') {
    return quote(value);
  }
  if (value.constructor === Date) {
    return `new Date("${value.toISOString()}")`;
  }
  return String(value);
}

function traverse(current: ServerValue): string | { referred: string } {
  // Skip primitives
  if (isPrimitive(current)) {
    return serializePrimitive(current);
  }
  // Get reference of the current object
  let refResult = REFS.get(current);
  if (refResult != null) {
    // Return the reference if already assigned a value
    if (ASSIGNED_REFS.has(refResult)) {
      return { referred: getRefParam(refResult) };
    }
    // Early sign (to prevent recursion)
    ASSIGNED_REFS.add(refResult);
  }
  // Transform Set
  if (constructorCheck<Set<ServerValue>>(current, Set)) {
    if (current.size) {
      const values: string[] = [];

      STACK.add(current);
      for (const value of current.keys()) {
        if (STACK.has(value)) {
          // This object received a ref, we must generate a setter ref
          if (refResult == null) {
            refResult = createRef(current);
          }
          // Received a ref, this might be a recursive ref, defer an assignment
          createSetAdd(refResult, getRefParam(createRef(value)));
        } else {
          const result = traverse(value);
          if (typeof result === 'string') {
            values.push(result);
          } else {
            values.push(result.referred);
          }
        }
      }
      STACK.delete(current);

      const value = values.length ? `new Set([${join(values, ',')}])` : EMPTY_SET;
      if (refResult != null) {
        return assignRef(refResult, value);
      }
      return value;
    }
    return EMPTY_SET;
  }
  if (constructorCheck<Map<ServerValue, ServerValue>>(current, Map)) {
    if (current.size) {
      const values: string[] = [];

      STACK.add(current);
      for (const [key, value] of current.entries()) {
        if (STACK.has(key)) {
          if (refResult == null) {
            refResult = createRef(current);
          }
          const keyRef = getRefParam(createRef(key));
          if (STACK.has(value)) {
            createMapSet(refResult, keyRef, getRefParam(createRef(value)));
          } else {
            const valueResult = traverse(value);
            if (typeof valueResult === 'string') {
              createMapSet(refResult, keyRef, valueResult);
            } else {
              createMapSet(refResult, keyRef, valueResult.referred);
            }
          }
        } else if (STACK.has(value)) {
          if (refResult == null) {
            refResult = createRef(current);
          }
          const valueRef = getRefParam(createRef(value));
          const keyResult = traverse(key);
          if (typeof keyResult === 'string') {
            createMapSet(refResult, keyResult, valueRef);
          } else {
            createMapSet(refResult, keyResult.referred, valueRef);
          }
        } else {
          const keyResult = traverse(key);
          const valueResult = traverse(value);
          if (typeof keyResult === 'string') {
            if (typeof valueResult === 'string') {
              values.push(`[${keyResult},${valueResult}]`);
            } else {
              values.push(`[${keyResult},${valueResult.referred}]`);
            }
          } else {
            if (typeof valueResult === 'string') {
              values.push(`[${keyResult.referred},${valueResult}]`);
            } else {
              values.push(`[${keyResult.referred},${valueResult.referred}]`);
            }
          }
        }
      }
      STACK.delete(current);
      const value = values.length ? `new Map([${join(values, ',')}])` : EMPTY_MAP;
      if (refResult != null) {
        return assignRef(refResult, value);
      }
      return value;
    }
    return EMPTY_MAP;
  }
  if (Array.isArray(current)) {
    if (current.length) {
      let values = '';

      STACK.add(current);
      forEach(current, (item, i) => {
        if (i in current) {
          if (STACK.has(item)) {
            if (refResult == null) {
              refResult = createRef(current);
            }
            createArrayAssign(refResult, i, getRefParam(createRef(item)));
            values += ',';
          } else {
            const result = traverse(item);
            if (typeof result === 'string') {
              values += result;
            } else {
              values += result.referred;
            }
            if (i < current.length - 1) {
              values += ',';
            }
          }
        } else {
          values += ',';
        }
      });
      STACK.delete(current);

      const value = `[${values}]`;
      if (refResult != null) {
        return assignRef(refResult, value);
      }
      return value;
    }
    return EMPTY_ARRAY;
  }
  if (current.constructor === Object) {
    const values: string[] = [];

    STACK.add(current);
    forEach(Object.entries(current), ([key, value]) => {
      const check = Number(key);
      // Test if key is a valid number or JS identifier
      // so that we don't have to serialize the key and wrap with brackets
      if (Number.isNaN(check) || /^([$A-Z_][0-9A-Z_$]*)$/i.test(key)) {
        if (STACK.has(value)) {
          if (refResult == null) {
            refResult = createRef(current);
          }
          createObjectIdentifierAssign(refResult, key, getRefParam(createRef(value)));
        } else {
          const result = traverse(value);
          if (typeof result === 'string') {
            values.push(`${key}:${result}`);
          } else {
            values.push(`${key}:${result.referred}`);
          }
        }
      } else if (STACK.has(value)) {
        if (refResult == null) {
          refResult = createRef(current);
        }
        createObjectStringAssign(refResult, key, getRefParam(createRef(value)));
      } else {
        const result = traverse(value);
        if (typeof result === 'string') {
          values.push(`${quote(key)}:${result}`);
        } else {
          values.push(`${quote(key)}:${result.referred}`);
        }
      }
    });
    STACK.delete(current);

    const value = `{${join(values, ',')}}`;
    if (refResult != null) {
      return assignRef(refResult, value);
    }
    return value;
  }
  throw new Error('Unserializable value');
}

export default function serialize(source: ServerValue) {
  if (isPrimitive(source)) {
    return serializePrimitive(source);
  }
  reset();
  // Lookup possible shared references
  lookupRefs(source);
  // Add shared references to refs
  for (const [key, value] of REF_COUNT.entries()) {
    if (value > 1) {
      createRef(key);
    }
  }
  // Get top-level serialization
  const result = traverse(source);
  if (typeof result !== 'string') {
    throw new Error('Unreachable error');
  }

  // Shared references detected
  if (VARS.length) {
    // Get (or create) a ref from the source
    const index = getRefParam(createRef(source));
    const patches = resolvePatches();
    const params = VARS.length > 1
      ? `(${join(VARS, ',')})`
      : VARS[0];
    // Source is probably already assigned
    if (result.startsWith(index)) {
      return `(${params}=>(${result},${patches}${index}))()`;
    }
    return `(${params}=>(${index}=${result},${patches}${index}))()`;
  }
  if (source.constructor === Object) {
    return `(${result})`;
  }
  return result;
}
