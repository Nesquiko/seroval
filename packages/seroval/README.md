# seroval

> Stringify JS values

[![NPM](https://img.shields.io/npm/v/seroval.svg)](https://www.npmjs.com/package/seroval) [![JavaScript Style Guide](https://badgen.net/badge/code%20style/airbnb/ff5a5f?icon=airbnb)](https://github.com/airbnb/javascript)

## Install

```bash
npm install --save seroval
```

```bash
yarn add seroval
```

```bash
pnpm add seroval
```

## Usage

```js
import seroval from 'seroval';

const object = {
  number: [Math.random(), -0, NaN, Infinity, -Infinity],
  string: ['hello world', '<script>Hello World</script>'],
  boolean: [true, false],
  null: null,
  undefined: undefined,
  bigint: 9007199254740991n,
  array: [,,,], // holes
  regexp: /[a-z0-9]+/i,
  date: new Date(),
  map: new Map([['hello', 'world']]),
  set: new Set(['hello', 'world']),
};

// self cyclic references
// recursive objects
object.self = object;
// recursive arrays
object.array.push(object.array);
// recursive maps
object.map.set('self', object.map);
// recursive sets
object.set.add(object.set);

// mutual cyclic references
object.array.push(object.map);
object.map.set('mutual', object.set);
object.set.add(object.array);

const result = seroval(object);
```

Output (as a string):

```js
((h,j,k,m)=>(h={number:[0.5906513825027921,-0,NaN,Infinity,-Infinity],string:["hello world","\x3Cscript>Hello World\x3C/script>"],boolean:[!0,!1],null:null,undefined:void 0,bigint:BigInt("9007199254740991"),array:j=[,,,,k=new Map([["hello","world"],["mutual",m=new Set(["hello","world"])]])],regexp:/[a-z0-9]+/i,date:new Date("2023-02-06T06:16:00.384Z"),map:k,set:m},j[3]=j,h.self=h,k.set("self",k),m.add(m).add(j),h))()

// Formatted for readability
((h, j, k, m) => (h = {
  number: [0.5906513825027921, -0, NaN, Infinity, -Infinity],
  string: ["hello world", "\x3Cscript>Hello World\x3C/script>"],
  boolean: [!0, !1],
  null: null,
  undefined: void 0,
  bigint: BigInt("9007199254740991"),
  array: j = [, , , , k = new Map([
    ["hello", "world"],
    ["mutual", m = new Set(["hello", "world"])]
  ])],
  regexp: /[a-z0-9]+/i,
  date: new Date("2023-02-06T06:16:00.384Z"),
  map: k,
  set: m
}, j[3] = j, h.self = h, k.set("self", k), m.add(m).add(j), h))()
```

### Mutual cyclic example

```js
import seroval from 'seroval';

const a = new Map([['name', 'a']]);
const b = new Map([['name', 'b']]);
const c = new Map([['name', 'c']]);
const d = new Map([['name', 'd']]);

c.set('left', a);
d.set('left', a);
c.set('right', b);
d.set('right', b);
a.set('children', [c, d]);
b.set('children', [c, d]);

const result = seroval({ a, b, c, d });
```

Output (as a string):

```js
((h,j,k,m,o,q)=>(q={a:h=new Map([["name","a"],["children",[j=new Map([["name","c"],["right",k=new Map([["name","b"],["children",o=[,m=new Map([["name","d"]])]]])]]),m]]]),b:k,c:j,d:m},o[0]=j,j.set("left",h),m.set("left",h).set("right",k),q))()

// Formatted
((h, j, k, m, o, q) => (q = {
  a: h = new Map([
    ["name", "a"],
    ["children", [j = new Map([
      ["name", "c"],
      ["right", k = new Map([
        ["name", "b"],
        ["children", o = [, m = new Map([
          ["name", "d"]
        ])]]
      ])]
    ]), m]]
  ]),
  b: k,
  c: j,
  d: m
}, o[0] = j, j.set("left", h), m.set("left", h).set("right", k), q))()
```

## Supports

The following values are the only values accepted by `seroval`:

- Exact values
  - `NaN`
  - `Infinity`
  - `-Infinity`
  - `-0`
- Primitives
  - `number`
  - `string`
  - `boolean`
  - `null`
  - `undefined`
  - `bigint`
- `Array` + holes
- `Object`
  - `RegExp`
  - `Date`
  - `Map`
  - `Set`
- Cyclic references (both self and mutual)

## License

MIT Â© [lxsmnsyc](https://github.com/lxsmnsyc)
