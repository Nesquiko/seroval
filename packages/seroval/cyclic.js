import seroval from 'seroval';


const a = new Map([['name', 'a']]);
const b = new Map([['name', 'b']]);
const c = new Map([['name', 'c']]);
const d = new Map([['name', 'd']]);

c.set('left', a);
d.set('left', a);
c.set('right', b);
d.set('right', b);
a.set('children', [c, d]);
b.set('children', [c, d]);

const result = seroval({ a, b, c, d });
console.log(result);

