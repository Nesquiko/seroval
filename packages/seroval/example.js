import seroval from 'seroval';

const object = {
  number: [Math.random(), -0, NaN, Infinity, -Infinity],
  string: ['hello world', '<script>Hello World</script>'],
  boolean: [true, false],
  null: null,
  undefined: undefined,
  bigint: 9007199254740991n,
  array: [,,,], // holes
  regexp: /[a-z0-9]+/i,
  date: new Date(),
  map: new Map([['hello', 'world']]),
  set: new Set(['hello', 'world']),
};

// self cyclic references
// recursive objects
object.self = object;
// recursive arrays
object.array.push(object.array);
// recursive maps
object.map.set('self', object.map);
// recursive sets
object.set.add(object.set);

// mutual cyclic references
object.array.push(object.map);
object.map.set('mutual', object.set);
object.set.add(object.array);

const result = seroval(object);
console.log(result);