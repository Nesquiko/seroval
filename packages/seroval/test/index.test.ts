/* eslint-disable no-eval */
import { describe, it, expect } from 'vitest';
import seroval, { ServerValue } from '../src';

describe('seroval', () => {
  it('supports booleans', () => {
    expect(seroval(true)).toBe('!0');
    expect(seroval(false)).toBe('!1');
  });
  it('supports numbers', () => {
    const value = Math.random();
    expect(seroval(value)).toBe(`${value}`);
    expect(seroval(NaN)).toBe('NaN');
    expect(seroval(Infinity)).toBe('Infinity');
    expect(seroval(-Infinity)).toBe('-Infinity');
    expect(seroval(-0)).toBe('-0');
  });
  it('supports strings', () => {
    expect(seroval('"hello"')).toMatchSnapshot();
    expect(seroval('<script></script>')).toMatchSnapshot();
    expect(eval(seroval('"hello"'))).toBe('"hello"');
    expect(eval(seroval('<script></script>'))).toBe('<script></script>');
  });
  it('supports bigint', () => {
    expect(seroval(9007199254740991n)).toMatchSnapshot();
    expect(eval(seroval(9007199254740991n))).toBe(9007199254740991n);
  });
  it('supports Arrays', () => {
    const example = [1, 2, 3];
    const result = seroval(example);
    expect(result).toMatchSnapshot();
    const back = eval(result) as number[];
    expect(back).toBeInstanceOf(Array);
    expect(back[0]).toBe(example[0]);
    expect(back[1]).toBe(example[1]);
    expect(back[2]).toBe(example[2]);
  });
  it('supports Objects', () => {
    const example = { hello: 'world' };
    const result = seroval(example);
    expect(result).toMatchSnapshot();
    const back = eval(result) as Record<string, string>;
    expect(back).toBeInstanceOf(Object);
    expect(back.hello).toBe(example.hello);
  });
  it('supports Set', () => {
    const example = new Set([1, 2, 3]);
    const result = seroval(example);
    expect(result).toMatchSnapshot();
    const back = eval(result) as Set<number>;
    expect(back).toBeInstanceOf(Set);
    expect(back.has(1)).toBe(example.has(1));
    expect(back.has(2)).toBe(example.has(2));
    expect(back.has(3)).toBe(example.has(3));
  });
  it('supports Map', () => {
    const example = new Map([[1, 2], [3, 4]]);
    const result = seroval(example);
    expect(result).toMatchSnapshot();
    const back = eval(result) as Map<number, number>;
    expect(back).toBeInstanceOf(Map);
    expect(back.get(1)).toBe(example.get(1));
    expect(back.get(3)).toBe(example.get(3));
  });
  it('supports Date', () => {
    const example = new Date();
    const result = eval(seroval(example)) as Date;
    expect(result).toBeInstanceOf(Date);
    expect(result.toISOString()).toBe(example.toISOString());
  });
  it('supports RegExp', () => {
    const example = /[a-z0-9]+/i;
    const result = seroval(example);
    expect(result).toMatchSnapshot();
    const back = eval(result) as RegExp;
    expect(back).toBeInstanceOf(RegExp);
    expect(String(back)).toBe(String(example));
  });
  it('supports array holes', () => {
    const example = new Array(10);
    const result = seroval(example);
    expect(result).toMatchSnapshot();
    const back = eval(result) as ServerValue[];
    expect(0 in back).toBeFalsy();
    expect(back[0]).toBe(undefined);
    expect(back.length).toBe(example.length);
  });
  describe('self cyclic references', () => {
    it('supports Arrays', () => {
      const example: ServerValue[] = [];
      example[0] = example;
      example[1] = example;
      const result = seroval(example);
      expect(result).toMatchSnapshot();
      const back = eval(result) as ServerValue[];
      expect(back[0]).toBe(back);
      expect(back[1]).toBe(back);
    });
    it('supports Objects', () => {
      const example: Record<string, ServerValue> = {};
      example.a = example;
      example.b = example;
      const result = seroval(example);
      expect(result).toMatchSnapshot();
      const back = eval(result) as Record<string, ServerValue>;
      expect(back.a).toBe(back);
      expect(back.b).toBe(back);
    });
    it('supports Maps', () => {
      const example: Map<ServerValue, ServerValue> = new Map();
      example.set(example, example);
      const result = seroval(example);
      expect(result).toMatchSnapshot();
      const back = eval(result) as Map<ServerValue, ServerValue>;
      expect(back.has(back)).toBe(true);
      expect(back.get(back)).toBe(back);
    });
    it('supports Sets', () => {
      const example: Set<ServerValue> = new Set();
      example.add(example);
      const result = seroval(example);
      expect(result).toMatchSnapshot();
      const back = eval(result) as Set<ServerValue>;
      expect(back.has(back)).toBe(true);
    });
  });
  describe('mutual cyclic references', () => {
    it('supports Arrays and Arrays', () => {
      const a: ServerValue[] = [];
      const b: ServerValue[] = [];
      a[0] = b;
      b[0] = a;
      const example = [a, b];
      const result = seroval(example);
      expect(result).toMatchSnapshot();
      const back = eval(result) as ServerValue[][];
      expect(back[0]).toBe(back[1][0]);
      expect(back[1]).toBe(back[0][0]);
    });
    it('supports Arrays and Objects', () => {
      const a: ServerValue[] = [];
      const b: Record<string, ServerValue> = {};
      a[0] = b;
      b[0] = a;
      const example = [a, b];
      const result = seroval(example);
      expect(result).toMatchSnapshot();
      const back = eval(result) as ServerValue[][];
      expect(back[0]).toBe(back[1][0]);
      expect(back[1]).toBe(back[0][0]);
    });
    it('supports Objects and Objects', () => {
      const a: Record<string, ServerValue> = {};
      const b: Record<string, ServerValue> = {};
      a[0] = b;
      b[0] = a;
      const example = [a, b];
      const result = seroval(example);
      expect(result).toMatchSnapshot();
      const back = eval(result) as ServerValue[][];
      expect(back[0]).toBe(back[1][0]);
      expect(back[1]).toBe(back[0][0]);
    });
  });
});
