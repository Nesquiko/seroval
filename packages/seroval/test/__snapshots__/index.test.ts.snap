// Vitest Snapshot v1

exports[`seroval > mutual cyclic references > supports Arrays and Arrays 1`] = `"((h,j,k)=>(k=[h=[j=[,]],j],j[0]=h,k))()"`;

exports[`seroval > mutual cyclic references > supports Arrays and Objects 1`] = `"((h,j,k)=>(k=[h=[j={}],j],j[\\"0\\"]=h,k))()"`;

exports[`seroval > mutual cyclic references > supports Objects and Objects 1`] = `"((h,j,k)=>(k=[h={\\"0\\":j={}},j],j[\\"0\\"]=h,k))()"`;

exports[`seroval > self cyclic references > supports Arrays 1`] = `"(h=>(h=[,,],h[1]=h[0]=h,h))()"`;

exports[`seroval > self cyclic references > supports Maps 1`] = `"(h=>(h=new Map,h.set(h,h),h))()"`;

exports[`seroval > self cyclic references > supports Objects 1`] = `"(h=>(h={},h.b=h.a=h,h))()"`;

exports[`seroval > self cyclic references > supports Sets 1`] = `"(h=>(h=new Set,h.add(h),h))()"`;

exports[`seroval > supports Arrays 1`] = `"[1,2,3]"`;

exports[`seroval > supports Map 1`] = `"new Map([[1,2],[3,4]])"`;

exports[`seroval > supports Objects 1`] = `"({hello:\\"world\\"})"`;

exports[`seroval > supports RegExp 1`] = `"/[a-z0-9]+/i"`;

exports[`seroval > supports Set 1`] = `"new Set([1,2,3])"`;

exports[`seroval > supports array holes 1`] = `"[,,,,,,,,,,]"`;

exports[`seroval > supports bigint 1`] = `"BigInt(\\"9007199254740991\\")"`;

exports[`seroval > supports strings 1`] = `"\\"\\\\\\"hello\\\\\\"\\""`;

exports[`seroval > supports strings 2`] = `"\\"\\\\x3Cscript>\\\\x3C/script>\\""`;
